Automated MMR signing key generation process

Michael-Hollister opened the pull request:
Automated signing key setup process described in https://docs.t2bot.io/matrix-media-repo/v1.3.5/installation/signing-key/

Homeserver backup key is kept alongside the merged signing key in case server administrator needs to revert.

spantaleev commented on roles/custom/matrix-media-repo/tasks/setup_install.yml:
```diff
@@ -77,6 +77,58 @@
+          --name={{ matrix_media_repo_identifier }}
+          --user={{ matrix_synapse_uid }}:{{ matrix_synapse_gid }}
+          --cap-drop=ALL
+          --mount type=bind,src={{ matrix_media_repo_config_path }},dst=/config
+          --mount type=bind,src={{ matrix_base_data_path }},dst=/matrix
+          --workdir='/config'
+          --entrypoint='combine_signing_keys'
+          {{ matrix_media_repo_docker_image }}
+          -format {{ matrix_homeserver_implementation }} -output {{ matrix_media_repo_homeserver_signing_key }}.merged {{ matrix_media_repo_homeserver_signing_key }} {{ matrix_media_repo_identifier }}.signing.key
+        creates: "{{ matrix_media_repo_homeserver_signing_key }}."
```
You're mounting the base path as `/matrix`, but then are accessing some file (`matrix_media_repo_homeserver_signing_key`).

If someone used a different path (e.g. `/xyz`), the file path in `matrix_media_repo_homeserver_signing_key` (which respects `matrix_base_data_path`) would be something like `/xyz/synapse/config/...`.

There's a mismatch been what's available in this container (the base path being exposed at `/matrix`) and what would be accessed (`/xyz/synapse/config/...`).

-----

The `creates` part of this Ansible command invocation seems off. Not sure what your intention was. Maybe it should have a `.merged` suffix?

-----

It's probably easiest if you do it like this:

```suggestion
          --mount type=bind,src={{ matrix_media_repo_config_path }},dst=/config
          --mount type=bind,src={{ matrix_media_repo_homeserver_signing_key | dirname }},dst=/homeserver-signing-key-dir,ro
          --workdir='/config'
          --entrypoint='combine_signing_keys'
          {{ matrix_media_repo_docker_image }}
          -format {{ matrix_homeserver_implementation }} -output /homeserver-signing-key-dir/{{ matrix_media_repo_homeserver_signing_key | basename }}.merged /homeserver-signing-key-dir/{{ matrix_media_repo_homeserver_signing_key | basename }} {{ matrix_media_repo_identifier }}.signing.key
        creates: "{{ matrix_media_repo_homeserver_signing_key }}."
```

spantaleev commented on group_vars/matrix_servers:
```diff
@@ -3642,6 +3642,11 @@ matrix_media_repo_systemd_required_services_list_auto: |
     ([devture_postgres_identifier ~ '.service'] if devture_postgres_enabled and matrix_media_repo_database_hostname == devture_postgres_connection_hostname else [])
   }}
 
+matrix_media_repo_generate_signing_key: "{{ matrix_homeserver_implementation == 'synapse' or matrix_homeserver_implementation == 'dendrite'}}"
+matrix_media_repo_homeserver_signing_key: "{{ matrix_media_repo_synapse_signing_key if matrix_homeserver_implementation == 'synapse' else (matrix_media_repo_dendrite_signing_key if matrix_homeserver_implementation == 'dendrite' else '') }}"
```
It seems like these 2 should have definitions in `defaults/main.yml` in the role, given that they are used within the role.

`matrix_media_repo_synapse_signing_key` and `matrix_media_repo_dendrite_signing_key` seem like helper variables used by the playbook, so they're OK to remain in `group_vars/matrix_servers` only.

spantaleev commented on roles/custom/matrix-media-repo/tasks/setup_install.yml:
```diff
@@ -77,6 +77,58 @@
+    path: "{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key"
+  register: matrix_media_repo_signing_key_stat
+
+- when: "matrix_media_repo_generate_signing_key | bool and not (matrix_media_repo_signing_key_stat.stat.exists | bool)"
+  block:
```
All the magic happens in this guard block, but.. The `when` will be satisfied early on (in the first sub-task) and there are more tasks after that.

This means that if the playbook dies after the first task creates the file and someone re-runs the playbook later on, we would never again enter this `when` block again. The job would effectively be half-done, without a chance to make it fully done except for figuring out what happened and deleting the `{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key` file.

-----

A potential solution is:

1. Continue guarding on ` {{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key` as you do now
2. Generate  `{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key.TEMPORARY` instead of  `{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key`
3. Work with `{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key.TEMPORARY` in the `block`
4. Move `{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key.TEMPORARY` to `{{ matrix_media_repo_config_path }}/{{ matrix_media_repo_identifier }}.signing.key` as a last step in this block

spantaleev reviewed (changes requested):
Other than what's suggested in the review (which I have not tested at all):

1. I wonder if this should be enabled by default, or if we should have some way for the user to control it. Right now, people can force-set `matrix_media_repo_generate_signing_key` to `false` to disable this behavior, but they must do it before it has already been done. Thoughts?

2. Given that the procedure is somewhat invasive, I wonder if we should mention it (and the [upstream docs page](https://docs.t2bot.io/matrix-media-repo/v1.3.5/installation/signing-key/) and playbook behavior - merging, backup, etc.) on our matrix-media-repo docs page?

3. Someone removing matrix-media-repo and deleting all its files (from `matrix_media_repo_base_path`) and re-enabling it would probably end up with a signing key which contains 3 keys (homeserver original key + 1st matrix-media-repo signing key + 2nd matrix-media-repo signing key). This may not be as bad as it sounds, I guess.

Michael-Hollister commented:
Thanks for the feedback, updated PR with your suggestions.

> 1. I wonder if this should be enabled by default, or if we should have some way for the user to control it. Right now, people can force-set `matrix_media_repo_generate_signing_key` to `false` to disable this behavior, but they must do it before it has already been done. Thoughts?
> 2. Given that the procedure is somewhat invasive, I wonder if we should mention it (and the [upstream docs page](https://docs.t2bot.io/matrix-media-repo/v1.3.5/installation/signing-key/) and playbook behavior - merging, backup, etc.) on our matrix-media-repo docs page?

Key generation is enabled by default and can be disabled by setting `matrix_media_repo_generate_signing_key` to false as you said. I think having the MMR docs page explaining the process would be sufficient. I've updated the MMR docs page explaining the process in more detail, and feel free to let me know if any further changes you'd like to see. 

> 3. Someone removing matrix-media-repo and deleting all its files (from `matrix_media_repo_base_path`) and re-enabling it would probably end up with a signing key which contains 3 keys (homeserver original key + 1st matrix-media-repo signing key + 2nd matrix-media-repo signing key). This may not be as bad as it sounds, I guess.

Yeah true, if a user manually deletes the MMR directory and re-runs the setup role, old keys will not be removed. If the user operates outside the playbook, I don't think this is a case we can easily protect against.

We could potentially auto remove old keys if the user runs the setup role with MMR disabled. However perhaps its best if left to the user to manually remove the key for the following reasons:
1. The playbook could restore from the `.backup` homeserver signing key. This could be destructive however since if the user modifies the signing key file after running the MMR setup role, those changes will be lost.
2. The playbook could instead search for the MMR key id in the homeserver signing key file before removal in the uninstall task. This could work, though a bit more code to maintain.
3. Its recommended to update your homeserver config when revoking/expiring old keys (`old_signing_keys` field for Synapse and `old_private_keys` for Dendrite). This would be harder to automate, and changes to the config would be lost when re-running the homeserver setup roles.

Feel free to discuss if you think otherwise.
